### ISSUE TITLE
[Extension Proposal] Generate AsyncApi-doc by scanning for annotations

### ISSUE THREAD

**Issue by ChMThiel** [quarkusio/quarkus#31238]
### Description

[AsyncApi](https://www.asyncapi.com/) specifies a way to describe asynchronous communication (like [OpenApi](https://www.openapis.org/) does for REST-Apis).

This extension will provide similar functionality for asynchronous communication, like OpenApi/Swagger does for REST-APIs:

- It will scan a Quarkus project for asynchronous communication using the microprofile annotations like` @Channel, @Incoming and @Outgoing`, generating an [AsyncApi-Object](https://github.com/asyncapi/jasyncapi) 
- The result can be filtered / post-processed
- The filtered-result will be a stored as file which can be 
  - downloaded as yaml or json
  - viewed directly in a html-page (like [here](https://studio.asyncapi.com/))

Interested in this extension, please +1 via the emoji/reaction feature of GitHub (top right).

### Repository name

quarkus-asyncapi-scanner

### Short description

generate AsyncApi by annotation-scanning

### Repository Homepage URL

https://quarkiverse.github.io/quarkiverse-docs/<REPOSITORY_NAME>/dev/

### Repository Topics

- quarkus-extension
- asyncApi
- kafka
...

### Team Members

- ChMThiel

**Comment by gastaldi** [comment]
How about making it a new module of https://github.com/quarkiverse/quarkus-asyncapi?

**Comment by gastaldi** [comment]
Oh wait, I think this is exactly what the extension in https://github.com/quarkiverse/quarkus-asyncapi does? Or am I missing something?

**Comment by ChMThiel** [comment]
I started with an attempt to add a second module to the extension https://github.com/quarkiverse/quarkus-asyncapi.
But soon i realized that this extension intended quite the opposite: 

The existing extension scans a given asyncapi-file and provides the infos found there as Microprofile-configuration (Kafka-topics, server-url, etc.). So it can be used by someone who has already an existing AsyncApi-Definition and wants to create a service according to it.

My quarkus-extension follows a "code first" approach (what's the case in the projects I'm involved). A annotation-scanner scans the existing code, combines it with the given mp-config and creates an AsyncApi-description with this infos.

It turned out, that was difficult to combine this two targets in one single extension (quarkus-asyncapi creates mp-configs, pretty much the same my extension requires as already given) - so i abandoned the idea to add a new module to the existing quarkus-asyncapi and create my own extension instead.
You can view my POC here https://github.com/ChMThiel/quarkus-asyncapi/tree/poc-annotation-scanner (which started in fact as a fork of quarkus-asyncapi)

**Comment by gastaldi** [comment]
The idea of having a new module is that it would be a different extension under the same repository. 

My proposal is to turn this:
```shell
.
├── deployment
├── docs
├── integration-tests
├── runtime
└── pom.xml

```

Into this (names may be different): 

```shell
.
├── quarkus-asyncapi-generator
│   ├── deployment
│   ├── docs
│   ├── integration-tests
│   └── runtime
├── quarkus-asyncapi-scanner
│   ├── deployment
│   ├── docs
│   ├── integration-tests
│   └── runtime
├── docs
└── pom.xml
```

But I'd like to know what's @fjtirado's opinion about this

**Comment by ChMThiel** [comment]
A week has passed... any suggestions @fjtirado ?

**Comment by ChMThiel** [comment]
Another week has passed... any progress?
@gastaldi let's create a separate project...

**Comment by fjtirado** [comment]
@ChMThiel Sorry for the huge delay, I did not realize the issue existed till today. 
First of all, happy to hear about this initiative. 
Let me try to propose somehting. 
I was planning to generate Java code from the async API document to add an example with uses that generate code to invoke the smallry channel automatically configured by the existing code in the extension. 
So, this Java code I was planning to generate can contain the annotations you were planning to scan to generate the async api  and this way we might  have a fully symetric annotation: async api file generator from annotated java code and annotated  java code  + smallrye config generator from async api definition files
So, I will kindly ask you to start defining these annotations in a module within the extenstion so once we have agreed on them, we can do the two things in parallel: develop code to create async definition from the annotations and develop code to generate java file with annotation from the async definitio. 
Does it sounds good to you?

**Comment by gastaldi** [comment]
I can add @ChMThiel as a committer in the quarkus-async-api repository if that helps, WDYT ?

**Comment by fjtirado** [comment]
@gastaldi sure

**Comment by gastaldi** [comment]
@fjtirado I created https://github.com/quarkiverse/quarkiverse-devops/pull/126 for your appreciation

**Comment by fjtirado** [comment]
If we are going for this, we need to find out an easy way for users to indicate if they are generating java from async or if they are generating ayns from java. 
I vote for adding a property that indicated that circumstance. Since Im terrible for naming, I just indicating the values: java (default), definition (the one that will need to be specified to generate definition from java)
The important thing is (one extension with discriminator property or two extension without property) that we share the annotations so we can arbitrarily go in one direction or the other.

**Comment by ChMThiel** [comment]
Right now we scan following annotations: 
* org.eclipse.microprofile.reactive.messaging.Channel
* org.eclipse.microprofile.reactive.messaging.Incoming
* org.eclipse.microprofile.reactive.messaging.Outgoing
To add a description to the channels, messages etc. we use the already existing OpenApi-Annotation
* org.eclipse.microprofile.openapi.annotations.media.Schema
That came in very handy because we use the same Objects for REST-APIs, too. So the @Schema-annotation are already present. (It's clear we are abusing the OpenApi-annotation a bit)
So we needed no additional, AsyncApi-dedicated annotations.

As mentioned in my earlier comment, i had problems to combine my feature into the Extension because my config-reading interfered with your config-writing.
For simplicity reasons i would prefer to separate the two extensions...

If you want to have a look at the current state of implementation: https://github.com/ChMThiel/quarkus-asyncapi/tree/poc-annotation-scanner

**Comment by fjtirado** [comment]
@ChMThiel Ok, you are using open api annotations. I was thinking to create a specific set of annotation for async (basically a clone of the openapi one) and put it into a spi repo that we can share to ensure we have simmetry between both generators. 
Yes, I think two separate extensions (quarkus-async-api and quarkus-java2Async-api) and one shared SPI extension (quarkus-asyncapi-annotations)  is the cleanest approach. 
There is little code to be shared between the two scanners-generators besides these annotations. And defining them, we do not "abuse" the open api annotations. Also Im not sure we have a "protocol" equivalent in openAPI, which is important for achieving simmetry between both extensions

**Comment by ChMThiel** [comment]
@gastaldi As discussed ... would you create a separate extension project?

**Comment by fjtirado** [comment]
@ChMThiel He is going to move things around withint the same repo so we can share the SPI annotation module, but in all other regards will be a different extension. 
I really believe to use OpenAPI schema annotaiton snould not be the long term strategy. It is cleaner and safer to create AsyncAPI specific annotations (which is as easy as forking the open api ones and move to a different package for now)

**Comment by fjtirado** [comment]
Let me clarify, I agree that the annotation should be called Schema, but I think it should be located within an async package. This way we allow both specs to evolve at a different pace. 
For you currently existing Java code will be as easy as changing the import statement.

**Comment by fjtirado** [comment]
Another example, For the channels, you are using smallry ones, but if you want to use the same channel name for subscribe and publish in the asyns api, you cannot, because smallrey channel names are globally unique. 
Therefore, I think we really need specific annotations for that and not reuse existing ones (you can still use the smallrye annotation in your code, but you will need to add the async api specific one)

**Comment by gastaldi** [comment]
Let me write this down, just so I understand the conclusion here:

- We'll have two extensions (or three if the plan is to share what's common) in the https://github.com/quarkiverse/quarkus-asyncapi - as proposed in https://github.com/quarkusio/quarkus/issues/31238#issuecomment-1435086249
- @ChMThiel will be granted push access to the https://github.com/quarkiverse/quarkus-asyncapi repository so he can push his work and release it when needed

Is that accurate?

**Comment by fjtirado** [comment]
@gastaldi yes, if you can create also the structure for the SPI annotations, I would really appreciate. We can always deleted it if at the end I agree that reusing the OpenAPI schema is a good approach.

**Comment by ChMThiel** [comment]
@gastaldi How should the project-structure look like? I'm not sure how to do it right... I would really appreciate if you could create the structures you have in mind. Then i can add/merge my code

**Comment by gastaldi** [comment]
@ChMThiel I've suggested it here: https://github.com/quarkusio/quarkus/issues/31238#issuecomment-1435086249

Basically it's a separate Maven module

**Comment by ChMThiel** [comment]
@gastaldi 
Thanks for your help on my first Quarkus-extension! 

But there's still one question: 
When i browse the extension [here](https://quarkus.io/extensions/) the quarkus-asyncapi-scanner is not listed (but the sister-extenson quarkus-asyncapi is). The docs of both extensions are show correctly [here](https://quarkiverse.github.io/quarkiverse-docs/quarkus-asyncapi/dev/).
What do i have to do to add the asyncap-annotation-scanner to the list of extensions?

**Comment by gastaldi** [comment]
@ChMThiel you need to add it to the quarkus-extension-catalog. See https://github.com/quarkiverse/quarkiverse/wiki/Checklist#make-your-extension-available-in-the-tooling

**Comment by gastaldi** [comment]
@ChMThiel I have created it and pushed here : https://github.com/quarkusio/quarkus-extension-catalog/commit/a76a6aa058adfcdfd8e0203af1dd1aa00a3278b4
